### PR TITLE
Improve connection handling after a disconnect

### PR DESCRIPTION
**Pull request by twyatt**
Unifies clean up of connection when no longer needed.

It's a bit of a callback dance, but there are a number of causes for disconnection when we have to be sure we've cleaned up the underlying `BluetoothGatt` (via `close`):

- `STATE_DISCONNECTED` from Android (connection dropped, e.g. device went out of range)
- `Peripheral.disconnect` (developer has explicitly requested a disconnect, in `finally` we `dispose` to be sure we clean up if disconnect process is cancelled)
- `Peripheral`'s parent `CoroutineScope` is cancelled
- `Peripheral.connect` is cancelled (e.g. user leaves current screen and connect was being performed via `viewModelScope`)

### CODE REVIEW

File: core/src/androidMain/kotlin/Observers.kt [hunk 1]
```diff
@@ -52,7 +52,13 @@ internal class Observers(
             }
         } finally {
             if (observers.decrementAndGet(characteristic) < 1) {
-                peripheral.stopNotifications(characteristic)
+                try {
+                    peripheral.stopNotifications(characteristic)
+                } catch (e: NotReadyException) {
+                    // Silently ignore as it is assumed that failure is due to connection drop, in which case Android
+                    // will clear the notifications.
+                    println("Stop notification failure ignored.")
```
Comment (Phoenix7351) [inline core/src/androidMain/kotlin/Observers.kt:60]: Do you have a logger available instead of println?

File: core/src/androidMain/kotlin/Peripheral.kt [hunk 2]
```diff
@@ -78,34 +77,39 @@ public class AndroidPeripheral internal constructor(
...
+        ) ?: throw ConnectionRejectedException()
+
+    /** Creates a connect [Job] that completes when connection is established, or failure occurs. */
     private fun createConnectJob(): Job = scope.launch(start = LAZY) {
         _ready.value = false
 
-        val connection =
-            bluetoothDevice.connect(context, _state) ?: throw ConnectionRejectedException()
-        _connection = connection
+        val connection = establishConnection().also { _connection = it }
```
Comment (Phoenix7351) [inline core/src/androidMain/kotlin/Peripheral.kt:91]: Was the switch to using "also" here stylistic? Or is there a more concrete reason to do this instead of just assigning _connection?

File: core/src/androidMain/kotlin/Peripheral.kt [hunk 3]
```diff
@@ -78,34 +77,39 @@ public class AndroidPeripheral internal constructor(
...
+        connection
+            .characteristicChanges
+            .onEach(observers.characteristicChanges::emit)
+            .launchIn(scope, start = UNDISPATCHED)
```
Comment (Phoenix7351) [inline core/src/androidMain/kotlin/Peripheral.kt:95]: Are you intentionally not catching the errors that could be thrown from this?

File: core/src/androidMain/kotlin/Peripheral.kt [hunk 4]
```diff
@@ -78,34 +77,39 @@ public class AndroidPeripheral internal constructor(
...
         }
 
         println("Ready")
         _ready.value = true
-    }.apply {
-        // fixme: Clear on disconnect (not at end of connect).
-        invokeOnCompletion { connectJob.value = null }
+    }
+
+    private fun dispose() {
```
Comment (Phoenix7351) [inline core/src/androidMain/kotlin/Peripheral.kt:110]: Should this function name be a little more verbose to point out that it's the connection being disposed? Looking at this from the outside, I would have guessed that this function would be cleaning up some state for Peripheral at large--not just the Peripheral's connection.

File: core/src/androidMain/kotlin/gatt/Callback.kt [hunk 5]
```diff
@@ -46,7 +46,7 @@ internal class Callback(
 ) : BluetoothGattCallback() {
 
     private var disconnectedAction: DisconnectedAction? = null
-    fun invokeOnDisconnected(action: () -> Unit) {
+    fun invokeOnDisconnected(action: DisconnectedAction) {
```
Comment (Phoenix7351) [inline core/src/androidMain/kotlin/gatt/Callback.kt:49]: Just curious, is DisconnectedAction just an alias for () -> Unit?

File: core/src/androidMain/kotlin/Peripheral.kt (hunk 2 above)
Comment (twyatt) [reply core/src/androidMain/kotlin/Peripheral.kt:91]: The `establishConnection` call shortened the line a bit, so make it so it was still reasonably readable when using `also`.

**tl;dr** yes, pure stylistic

File: core/src/androidMain/kotlin/Observers.kt (hunk 1 above)
Comment (twyatt) [reply core/src/androidMain/kotlin/Observers.kt:60]: Tracking via #6.
https://github.com/JuulLabs/tuulbox/pull/22 might provide the logging capabilities we're looking for. 😉

File: core/src/androidMain/kotlin/Peripheral.kt (hunk 3 above)
Comment (twyatt) [reply core/src/androidMain/kotlin/Peripheral.kt:95]: Correct.

Before this PR, upstream the `Channel` was being closed with an exception, only to have it catch here as a signal to stop collecting.
Instead, just performing a standard close (without exception) upstream so that this stops collection gracefully.

File: core/src/androidMain/kotlin/gatt/Callback.kt (hunk 5 above)
Comment (twyatt) [reply core/src/androidMain/kotlin/gatt/Callback.kt:49]: Yup

https://github.com/JuulLabs/kable/blob/226102b47e271a434ed865e72c72053e35a0c8c1/core/src/androidMain/kotlin/gatt/Callback.kt#L37

File: core/src/androidMain/kotlin/Peripheral.kt (hunk 4 above)
Comment (twyatt) [reply core/src/androidMain/kotlin/Peripheral.kt:110]: Technically, the `Connection` object internally holds either a dispatcher or thread that needs to be cleaned up (and is done via callback from the `Connection.close`). So maybe `Connection.close` isn't verbose enough? Happy to change any names that make things more readable, not sure what to change it to though. Also not as important for `private` functions. 🤷

**Review by Phoenix7351 — APPROVED** [review]

**Review by sdonn3 — APPROVED** [review]
lgtm
